docker/docker_client_test.go

```diff
@@ -187,3 +188,143 @@ func TestUserAgent(t *testing.T) {
 		}
 	}
 }
+
+func TestNeedsRetryOnError(t *testing.T) {
+	needsRetry, _ := needsRetryWithUpdatedScope(errors.New("generic"), nil)
+	if needsRetry {
+		t.Fatal("Got needRetry for a connection that included an error")
+	}
+}
+
+var registrySuseComResp = http.Response{
+	Status:     "401 Unauthorized",
+	StatusCode: http.StatusUnauthorized,
+	Proto:      "HTTP/1.1",
+	ProtoMajor: 1,
+	ProtoMinor: 1,
+	Header: map[string][]string{
+		"Content-Length":                  {"145"},
+		"Content-Type":                    {"application/json"},
+		"Date":                            {"Fri, 26 Aug 2022 08:03:13 GMT"},
+		"Docker-Distribution-Api-Version": {"registry/2.0"},
+		// "Www-Authenticate":                {`Bearer realm="https://registry.suse.com/auth",service="SUSE Linux Docker Registry",scope="registry:catalog:*",error="insufficient_scope"`},
+		"X-Content-Type-Options": {"nosniff"},
+	},
+	Request: nil,
+}
+
+func TestNeedsRetryOnInsuficientScope(t *testing.T) {
+	resp := registrySuseComResp
+	resp.Header["Www-Authenticate"] = []string{
+		`Bearer realm="https://registry.suse.com/auth",service="SUSE Linux Docker Registry",scope="registry:catalog:*",error="insufficient_scope"`,
+	}
+	expectedScope := authScope{
+		resourceType: "registry",
+		remoteName:   "catalog",
+		actions:      "*",
+	}
+
+	needsRetry, scope := needsRetryWithUpdatedScope(nil, &resp)
+
+	if !needsRetry {
+		t.Fatal("Expected needing to retry")
+	}
+
+	if expectedScope != *scope {
```

Comment:
For _future_ reference (absolutely no need to change this PR, please don’t): Most tests in c/image use `"github.com/stretchr/testify/assert"` / `"github.com/stretchr/testify/require"`, which makes these kinds of tests a bit less tedious to write.

This does work perfectly fine, of course.